xds/internal/xdsclient/authority.go

```diff
@@ -159,14 +159,27 @@ func (a *authority) handleResourceUpdate(resourceUpdate transport.ResourceUpdate
 		ServerConfig:    a.serverCfg,
 	}
 	updates, md, err := decodeAllResources(opts, rType, resourceUpdate)
-	a.updateResourceStateAndScheduleCallbacks(rType, updates, md)
+	a.updateResourceStateAndScheduleCallbacks(rType, updates, md, fc)
 	return err
 }
 
-func (a *authority) updateResourceStateAndScheduleCallbacks(rType xdsresource.Type, updates map[string]resourceDataErrTuple, md xdsresource.UpdateMetadata) {
+func (a *authority) updateResourceStateAndScheduleCallbacks(rType xdsresource.Type, updates map[string]resourceDataErrTuple, md xdsresource.UpdateMetadata, fc *transport.ADSFlowControl) {
 	a.resourcesMu.Lock()
 	defer a.resourcesMu.Unlock()
 
+	// We build a list of callback funcs to invoke, and invoke them at the end
+	// of this method instead of inline (when handling the update for a
+	// particular resource), because we want to make sure that all calls to
+	// `fc.Add` happen before any callbacks are invoked. This will ensure that
+	// the next read is never attempted before all callbacks are invoked, and
+	// the watchers have processed the update.
```

Comment:
are invoked which means the watchers have processed the update or this function failed to schedule on serializer.

Reply:
If we didn't do this, and instead invoked the watcher callbacks inline like this when handling an update (and similarly for the error and resource-not-found cases)

```golang
fc.Add()
a.serializer.ScheduleOr(watcher.OnUpdate(resource, fc), fc.OnDone)
```

it is possible that `fc.Add()` increments the flow-control, and before the other resources which are part of this update are handled here, the serializer could schedule the callback, and the watcher could end up processing the update and invoke `fc.onDone`. This would result in the ADS stream to call `Recv` before actually every watcher has handled the update.

Does this make sense?

Reply:
Yeah I get it. I was just suggesting adding the second case, because if this function fails to schedule the callback fc.onDone gets invoked here which also falls within the language you described (can't intersplice because might trigger the next iteration early).

Reply:
Alright sounds good.